### PR TITLE
Allow adding unique userdata to instances

### PR DESCRIPTION
**Pull request by LorenDB**
I did this because I'm contributing to an app that uses user profiles and needs to have only one instance of each profile.

By the way, I tried to follow the coding style used in SingleApplication.

### CODE REVIEW

**Comment by itay-grudev** [comment]
What is the use case for this?

**Comment by LorenDB** [comment]
I'm working on github.com/Nheko-reborn/nheko. I was trying to add profiles to nheko, but I wanted a way to only block windows from opening if there was already a window open with the profile that the user was trying to load. So in my case, I can have the default profile in one window running the latest stable release, the same account in another profile "git" in a window built on the latest code,  but when I try opening another window with the git profile, it raises the first git window.

**Comment by itay-grudev** [comment]
But your code adds the ability to set it on a instance but doesn't share it in in any way with the rest of the instances. To achieve your goal you would need to save this in the shared memory.

**Comment by LorenDB** [comment]
> But your code adds the ability to set it on a instance but doesn't share it in in any way with the rest of the instances. To achieve your goal you would need to save this in the shared memory.

I'm not sure what you're getting at.

Basically, I'm using the userdata to add to the data that is used to generate the key for the QSharedMemory. In my opinion, that should be all that is necessary.

**Comment by itay-grudev** [comment]
@LorenDB You are correct. I missed on line when I was doing the review last night.

**Review by itay-grudev — CHANGES_REQUESTED** [review]
Would you also be kind enough to add some documentation in the README?

File: singleapplication.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@
  * @param options Optional flags to toggle specific behaviour
  * @param timeout Maximum time blocking functions are allowed during app load
  */
-SingleApplication::SingleApplication( int &argc, char *argv[], bool allowSecondary, Options options, int timeout )
+SingleApplication::SingleApplication( int &argc, char *argv[], bool allowSecondary, Options options, QString userData, int timeout )
```
[the remark this answers is not in the document]
Comment (LorenDB) [reply singleapplication.cpp:39]: Sure

**Comment by LorenDB** [comment]
Just pushed changes. I did remove the setAppData function from SingleApplication because there's not really any way set up to modify the QSharedMemory key after new user data is set. I also made the naming agree (it's now user data everywhere).

**Comment by itay-grudev** [comment]
Ok. I like it.
@jonaski Do you mind reviewing it as well?

**Review by itay-grudev — APPROVED** [review]

**Comment by itay-grudev** [comment]
There is a false positive test failure in the Qt 6 build specs, but this is not related. I think it's a bug elsewhere because the tests started failing without the code having been changed.

File: singleapplication.cpp [hunk 2]
```diff
@@ -51,6 +51,10 @@ SingleApplication::SingleApplication( int &argc, char *argv[], bool allowSeconda
     // Store the current mode of the program
     d->options = options;
 
+    // Add any unique user data
+    if ( ! userData.isNull() )
```
Comment (jonaski) [inline singleapplication.cpp:55]: Shouldn't this be isEmpty instead?

File: singleapplication.h [hunk 3]
```diff
@@ -85,7 +85,7 @@ class SingleApplication : public QAPPLICATION_CLASS
      * Usually 4*timeout would be the worst case (fail) scenario.
      * @see See the corresponding QAPPLICATION_CLASS constructor for reference
      */
-    explicit SingleApplication( int &argc, char *argv[], bool allowSecondary = false, Options options = Mode::User, int timeout = 1000 );
+    explicit SingleApplication( int &argc, char *argv[], bool allowSecondary = false, Options options = Mode::User, int timeout = 1000, QString userData = "" );
```
Comment (jonaski) [inline singleapplication.h:88]: QString() instead of ""?

**Comment by jonaski** [comment]
@itay-grudev I added 2 comments, let me know if you agree.

File: singleapplication.cpp (hunk 2 above)
Comment (LorenDB) [reply singleapplication.cpp:55]: Hmm, you're right.

**Comment by LorenDB** [comment]
> @itay-grudev I added 2 comments, let me know if you agree.

I just updated to use those changes, I think they are both valid.

File: README.md [hunk 4]
```diff
@@ -154,7 +154,7 @@ API
 ### Members
 
 ```cpp
-SingleApplication::SingleApplication( int &argc, char *argv[], bool allowSecondary = false, Options options = Mode::User, int timeout = 100 )
+SingleApplication::SingleApplication( int &argc, char *argv[], bool allowSecondary = false, Options options = Mode::User, int timeout = 100, QString userData = "" )
```
Comment (jonaski) [inline README.md:157]: QString() here to match the constructor

Comment (LorenDB) [reply README.md]: Done

**Review by jonaski — APPROVED** [review]

**Comment by itay-grudev** [comment]
Thank you @LorenDB. I'll add a changelog entry and credit you as Loren Burkholder, with your permission.

**Comment by LorenDB** [comment]
Go right ahead!

**Pull request merged**